java-checks/src/main/java/org/sonar/java/checks/spring/SpelExpressionCheck.java

```diff
@@ -0,0 +1,246 @@
+    "[a-zA-Z_]\\w*+(\\[\\d++])*+(\\.[a-zA-Z_]\\w*+(\\[\\d++])*+)*+"
+  );
+
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.CLASS, Tree.Kind.INTERFACE, Tree.Kind.ANNOTATION_TYPE);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    getClassAndMemberAnnotations((ClassTree) tree)
+      .filter(SpelExpressionCheck::isSpringAnnotation)
+      .forEach(this::checkSpringAnnotationArguments);
+  }
+
+  private static Stream<AnnotationTree> getClassAndMemberAnnotations(ClassTree cls) {
+    return Stream.concat(
+      Stream.of(cls.modifiers().annotations()),
+      cls.members().stream().map(SpelExpressionCheck::getMemberAnnotations)
+    ).flatMap(Collection::stream);
+  }
+
+  private static List<AnnotationTree> getMemberAnnotations(Tree member) {
+    if (member.is(Tree.Kind.METHOD)) {
+      return ((MethodTree) member).modifiers().annotations();
+    } else if (member.is(Tree.Kind.VARIABLE)) {
+      return ((VariableTree) member).modifiers().annotations();
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  private static boolean isSpringAnnotation(AnnotationTree annotation) {
+    return annotation.symbolType().fullyQualifiedName().startsWith(SPRING_PREFIX);
+  }
+
+  private void checkSpringAnnotationArguments(AnnotationTree annotation) {
+    annotation.arguments().stream()
+      .filter(argument -> argument.is(Tree.Kind.STRING_LITERAL))
+      .forEach(argument -> checkSpringAnnotationStringArgument((LiteralTree) argument));
+  }
+
+  private void checkSpringAnnotationStringArgument(LiteralTree argument) {
+    try {
+      var argValue = argument.value();
+      var argContents = argValue.substring(1, argValue.length() - 1);
+      parseStringContents(argContents);
+    } catch (SyntaxError e) {
+      reportIssue(argument, e);
+    }
+  }
+
+  private void reportIssue(LiteralTree stringToken, SyntaxError error) {
+    var tokenStart = Position.startOf(stringToken);
+    var textSpan = new AnalyzerMessage.TextSpan(
+      tokenStart.line(),
+      tokenStart.columnOffset() + error.rangeStart,
+      tokenStart.line(),
+      tokenStart.columnOffset() + error.rangeEnd
+    );
+
+    var analyzerMessage = new AnalyzerMessage(this, context.getInputFile(), textSpan, error.getMessage(), 0);
+    ((DefaultJavaFileScannerContext) context).reportIssue(analyzerMessage);
+  }
+
+  private static void parseStringContents(String content) {
+    var i = 0;
+    while (i < content.length()) {
+      var c = content.charAt(i);
+      switch (c) {
+        case '$':
+          i = parseDelimitersAndContents(content, i + 1, i + 1, SpelExpressionCheck::checkValidPropertyPlaceholder);
+          break;
+        case '#':
+          i = parseDelimitersAndContents(content, i + 1, i + 1, SpelExpressionCheck::checkValidSpelExpression);
+          break;
+        default:
+          i++;
+          break;
+      }
+    }
+  }
+
+  private static int parseDelimitersAndContents(
```

Comment:
With 2 parameters of the same type with very similar names, you probably want to add some Javadoc to explain why `startIndex` is sometimes different from `rangeStart`

Reply:
Is there a relation between startIndex and rangeStart?

Reply:
Added a doc for the method + gave the parameter a more proper name (also in the other methods). Fixed.